### PR TITLE
Add to Thread default_options

### PR DESCRIPTION
**Pull request by rhearnorth**

### CODE REVIEW

File: lib/bigcommerce_api/base.rb [hunk 1]
```diff
@@ -1,30 +1,36 @@
...
       # if we're using Oauth, we're probably grabbing :store_hash
       # accept :store_url for legacy purposes
       if params[:store_url]
-        self.class.base_uri(params[:store_url] + '/api/v2/')
+        session_options[:base_uri] = "#{params[:store_url]}/api/v2/"
       else
-        self.class.base_uri("https://api.bigcommerce.com/stores/#{params[:store_hash]}/v2/")
+        session_options[:base_uri] = "https://api.bigcommerce.com/stores/#{params[:store_hash]}/v2"
       end
+      Thread.current["BigcommerceAPI"] ||= session_options
```
Comment (rhearnorth) [inline lib/bigcommerce_api/base.rb]: If I use `Thread.current["BigcommerceAPI#{self.__id__}"]` 
this id `id` an instance id and when 
```ruby 
    def self.default_options
      Thread.current["BigcommerceAPI#{self.__id__}"] ||= {}
    end
```

**Comment by bradleypriest** [comment]
Why doesn't just defining default_options work?

**Comment by bradleypriest** [comment]
I've added a failing test here https://github.com/lcpriest/bigcommerce_api/compare/master...threadsafety

File: lib/bigcommerce_api/base.rb (hunk 1 above)
Comment (bradleypriest) [inline lib/bigcommerce_api/base.rb]: This probably should just be =

File: lib/bigcommerce_api/base.rb [hunk 2]
```diff
@@ -1,30 +1,36 @@
...
-        self.class.headers 'X-Auth-Token' => params[:access_token]
+        session_options[:headers]['X-Auth-Client'] = params[:client_id]
+        session_options[:headers]['X-Auth-Token']  = params[:access_token]
       end
 
       # if we're using Oauth, we're probably grabbing :store_hash
       # accept :store_url for legacy purposes
       if params[:store_url]
-        self.class.base_uri(params[:store_url] + '/api/v2/')
+        session_options[:base_uri] = "#{params[:store_url]}/api/v2/"
```
Comment (bradleypriest) [inline lib/bigcommerce_api/base.rb]: def base_uri actually has a transform step, can we include that as well

Comment (rhearnorth) [inline lib/bigcommerce_api/base.rb]: @bradleypriest sorry I don't understand 
https://github.com/jnunemaker/httparty/blob/f8cbf77cd1122ed0c97fa92081555ac7155948e7/lib/httparty.rb#L93

Comment (bradleypriest) [inline lib/bigcommerce_api/base.rb]: `session_options[:base_uri] = HTTParty.normalize_base_uri("#{params[:store_url]}/api/v2/")`

File: lib/bigcommerce_api/base.rb [hunk 3]
```diff
@@ -1,30 +1,38 @@
...
-        self.class.base_uri("https://api.bigcommerce.com/stores/#{params[:store_hash]}/v2/")
-      end
+      session_options[:base_uri] = if params[:store_url]
+                                     "#{params[:store_url]}/api/v2"
+                                   else
+                                     "https://api.bigcommerce.com/stores/#{params[:store_hash]}/v2"
+                                   end
+      session_options[:base_uri] = HTTParty.normalize_base_uri(session_options[:base_uri])
+
+      Thread.current["BigcommerceAPI"] ||= session_options
```
Comment (bradleypriest) [inline lib/bigcommerce_api/base.rb]: I'm still pretty sure this should be `=` instead of `||=`

Comment (rhearnorth) [inline lib/bigcommerce_api/base.rb]: @bradleypriest updated

File: lib/bigcommerce_api/base.rb [hunk 4]
```diff
@@ -1,30 +1,38 @@
...
-    headers 'Accept' => "application/json"
-    headers 'Content-Type' => "application/json"
+
+    def self.default_options
+      Thread.current["BigcommerceAPI"] ||= {}
+    end
 
     def initialize(params={})
       # for the time being, accept old school API params
+      session_options = { parser: HTTParty::Parser, format: :json, headers: { "Accept" => "application/json", "Content-Type" => "application/json" } }
```
Comment (bradleypriest) [inline lib/bigcommerce_api/base.rb]: Can you move this up to default_options

```ruby
def self.default_options
  Thread.current["BigcommerceAPI"] ||= { 
    parser: HTTParty::Parser, 
    format: :json, 
    headers: { "Accept" => "application/json", "Content-Type" => "application/json" } 
  }
end
```

Comment (bradleypriest) [inline lib/bigcommerce_api/base.rb]: And then use `default_options.merge!` at the bottom

**Comment by bradleypriest** [comment]
Just added a more relevant test. Yay, it still works!

File: lib/bigcommerce_api/base.rb (hunk 4 above)
Comment (rhearnorth) [inline lib/bigcommerce_api/base.rb]: can I do this
```ruby
    def initialize(params={})
      # for the time being, accept old school API params
      session_options = self.class.default_options
      if params[:username] and params[:api_key]
          session_options[:username] = params[:username]
          session_options[:password] = params[:api_key]
        # default to Oauth
      else
        session_options[:headers]['X-Auth-Client'] = params[:client_id]
        session_options[:headers]['X-Auth-Token']  = params[:access_token]
      end

      # if we're using Oauth, we're probably grabbing :store_hash
      # accept :store_url for legacy purposes
      session_options[:base_uri] = if params[:store_url]
                                     "#{params[:store_url]}/api/v2"
                                   else
                                     "https://api.bigcommerce.com/stores/#{params[:store_hash]}/v2"
                                   end
      session_options[:base_uri] = HTTParty.normalize_base_uri(session_options[:base_uri])

      Thread.current["BigcommerceAPI"] = session_options
    end
```

Comment (bradleypriest) [inline lib/bigcommerce_api/base.rb]: LGTM

**Comment by rhearnorth** [comment]
@bradleypriest I merged 2 tests into one commit, ready for you to push the button

File: lib/bigcommerce_api/base.rb [hunk 5]
```diff
@@ -1,30 +1,42 @@
+require 'thread'
```
Comment (gherry) [inline lib/bigcommerce_api/base.rb]: If I'm not mistaken, you don't actually need to require `thread` here if you only use it to get `Thread.current`

**Comment by gherry** [comment]
Yeah, this should do it. LGTM

**Comment by bradleypriest** [comment]
Just reran the test without the modifications and it broke, so we are definitely fixing something here

**Comment by rhearnorth** [comment]
Aye Aye Sir

**Comment by pmdeazeta** [comment]
:+1:

**Pull request merged**